Add notes for accessibility testing to development documentation (#13014)
PrairieLearn/PrairieLearn: docs/dev-guide/guide.md

Reviewer: The changes here focus on keyboard accessibility, which is only part of the equation. We need to adhere to at least WCAG 2.1 AA, which has a lot more things to watch out for, including color contrast (we got dinged on this during third-party review in the past).

These docs shouldn't be comprehensive, but they should focus on the things that are often missed, much like the [instructor question accessibility guidelines](https://prairielearn.readthedocs.io/en/latest/question/accessibility/#semantic-html) do.

Author: It's probably better if you make these tweaks directly. I have no experience with WCAG 2.1 AA outside of "ensure contrast is 3:1 between elements"

Reviewer: Let's continue discussion about those in the other comments I just left.

- Buttons should use the `<button>` element when they take actions and the `<a>` element when they are simply links to other pages. We should not use `<a role="button">` to fake a button element. Buttons that do not submit a form should always start with `<button type="button" class="btn ...">`, where `type="button"` specifies that they don't submit.

## HTML accessibility

> If you are adding anything more complex than a basic form page, the built-in accessibility checks are likely not enough for checking accessibility. You should use [voiceover (macOS)] or NVDA (windows) to test the page. Some common things to check for:

Reviewer: ```suggestion
If you are adding anything more complex than a basic form page, the automated accessibility checks are likely not enough for checking accessibility. You should use [VoiceOver (macOS)](https://support.apple.com/guide/voiceover/welcome/mac) or [NVDA (Windows)](https://www.nvaccess.org/download/) to test the page. Some common things to check for:
```

- Are elements announced correctly?
- Can you navigate the page using only the keyboard?
  - tab to get to the next focusable element, enter to activate it
  - arrow keys work appropriately

> - are the focus indicators visible?

Reviewer: ```suggestion
- Are the focus indicators visible?
```

Example bugs from work on the Students table:

- Can you navigate the table with arrow keys?
- Can you resize the table columns with the arrow keys?
- Can you activate a modal with enter?

> - Are the voiceover descriptions concise and accurate? Are there voiceover descriptions for all interactive elements (e.g. the resize handle)?

Reviewer: ```suggestion
- Are the announced descriptions concise and accurate? Are there announced descriptions for all interactive elements (e.g. the resize handle)?
```

- Does closing the modal/tabbing out of the modal retain the focused position?

> - Does selecting an option in a dropdown retain it's position after a re-render (e.g. in React)?

Reviewer: ```suggestion
- Does selecting an option in a dropdown retain its position after a re-render (e.g. in React)?
```

I don't actually understand what this is saying, can you clarify? What is "its position" with regards to a dropdown option?

Author: I edited the suggestion to be more clear.

- Buttons should use the `<button>` element when they take actions and the `<a>` element when they are simply links to other pages. We should not use `<a role="button">` to fake a button element. Buttons that do not submit a form should always start with `<button type="button" class="btn ...">`, where `type="button"` specifies that they don't submit.

## HTML accessibility

If you are adding anything more complex than a basic form page, the automated accessibility checks are likely not enough for checking accessibility. You should use [VoiceOver (macOS)](https://support.apple.com/guide/voiceover/welcome/mac) or [NVDA (Windows)](https://www.nvaccess.org/download/) to test the page. All our pages must conform to the [Web Content Accessibility Guidelines (WCAG) 2.1 AA standard](https://www.w3.org/TR/WCAG21/). Some common things to check for:

- Are elements announced correctly?
  - Do you have appropriate `aria-label` / `alt` attributes?
- Can you navigate the page using only the keyboard?
  - tab to get to the next focusable element, enter to activate it
  - arrow keys work appropriately
- Are the focus indicators visible?
- Is the page layout logical and easy to understand?

> 

Reviewer: ```suggestion
```

> - Will users with visual impairments be able to use the page? (More generally, we must follow)

Reviewer: ```suggestion
- Will users with visual impairments be able to use the page?
```

  - Do you have at least a 3:1 contrast ratio between text and background colors?
  - Do you have appropriate spacing between elements?

> 

Reviewer: ```suggestion
```

- Can you navigate the table with arrow keys?

> - Can you resize the table columns with the arrow keys?

Reviewer: What is "the table"?

Author: https://github.com/PrairieLearn/PrairieLearn/pull/13014#discussion_r2400042098

> - Can you activate a modal with enter?

Reviewer: Does "activate" mean "open"? I don't know why this is calling out modals specifically, it seems to me like this is a subset of "Can you navigate the page using only the keyboard" - buttons (regardless of whether or not they open a modal) should be triggerable with the keyboard.

> - Are the announced descriptions concise and accurate? Are there announced descriptions for all interactive elements (e.g. the resize handle)?

Reviewer: Shouldn't this be a subset of "are elements announced correctly"?

What is "the resize handle" in a generic context?

> - Does closing the modal/tabbing out of the modal retain the focused position?

Reviewer: When can one tab out of a modal; aren't they focus traps?

Author: I was thinking of this 
<img width="1150" height="638" alt="CleanShot 2025-10-02 at 17 00 23@2x" src="https://github.com/user-attachments/assets/4809a5f2-1aab-4e3e-ae91-77fa32c43c74" />

> - Is the same item still focused in a dropdown after an action that would cause a re-render of the dropdown?

Reviewer: I don't understand what this is saying. In a reply to this comment, can you point to a relevant code diff relating to this?

Author: This was specifically a problem in this component. When you froze a column, it would rerender, losing your original focused row position.

<img width="470" height="670" alt="CleanShot 2025-10-02 at 16 59 20@2x" src="https://github.com/user-attachments/assets/f29ea510-6d7b-478e-9694-35dd451e4bcd" />

> - Are you using the proper ARIA role on menus/toolbars? You should check this with the official role list.

Reviewer: If there's an "official role list", can you link to it?

Author: the mozilla list

## HTML accessibility

If you are adding anything more complex than a basic form page, the automated accessibility checks are likely not enough for checking accessibility. You should use [VoiceOver (macOS)](https://support.apple.com/guide/voiceover/welcome/mac) or [NVDA (Windows)](https://www.nvaccess.org/download/) to test the page. Some common things to check for:
If you are adding anything more complex than a basic form page, the automated accessibility checks are likely not enough for checking accessibility. You should use [VoiceOver (macOS)](https://support.apple.com/guide/voiceover/welcome/mac) or [NVDA (Windows)](https://www.nvaccess.org/download/) to test the page. All our pages must conform to the [Web Content Accessibility Guidelines (WCAG) 2.1 AA standard](https://www.w3.org/TR/WCAG21/). Some common things to check for:

- Are elements announced correctly?
  - Do you have appropriate `aria-label` / `alt` attributes?
- Can you navigate the page using only the keyboard?
  - tab to get to the next focusable element, enter to activate it
  - arrow keys work appropriately
- Are the focus indicators visible?
- Is the page layout logical and easy to understand?

> Example bugs from work on the Students table:

Author: It looks like there was a deleted line here. This was supposed to delineate some concrete examples of things I have done in the past.
